lib/dev-middleware.js

```diff
@@ -92,6 +92,20 @@ module.exports = function (app, config, options) {
                         }
                     }
 
+                    if (options.virtualWrite) {
+                        let vw_root = path.normalize(path.resolve(options.virtualWrite))
+                            .replace(process.cwd(), '')
+                            .split(path.sep)
+                            .filter(s => s !== '');
+
+                        fileName = path.normalize(path.resolve(configs[i].output.dir, fileName))
+                            .replace(process.cwd(), '')
+                            .split(path.sep)
+                            .filter(s => s !== '')
+                            .filter((s, i) => s !== vw_root[i])
```

Comment:
If I read the code correctly, I don't think we should drop partial "virtual path" matches from URLs.

In fact, I even think file names that don't have an exact match with the provided "virtual path" prefix should not be served. That's what a web server would do, and it also avoids possible URL conflicts since then there's a 1:1 mapping of filenames to URLs.

Should rather be something like:

~~~js
  if (fileName.startsWith(baseUrl)) {
    fileName = fileName.slice(virtualWrite.length)
  } else {
     return // don't serve the file ¯\_(ツ)_/¯
  }
~~~

Reply:
I'm not sure I understand what you mean. At the moment the following will happen with the following config:

```
--virtual-write dist

{
    dir: 'dist/app',
    entryFileNames: 'esm/app.js'
}
```

* Generated bundle will have ```esm/app.js``` as per normal.
* We prepend ```output.dir``` to the file URL: ```dist/app/esm/app.js```
* We then remove the  value of virtual write ```dist``` from the URL, this gives: ```app/esm/app.js```.
* This gets applied to ```files``` which contains all of the files that are accessible by the dev server.
* If you request ```esm/app.js``` or ```dist/app/esm/app.js```, it will fail. Only ```app/esm/app.js``` will respond.

Reply:
My concern is, for example, if you have `['dist', 'app1', 'build']` in `vw_root`, wouldn't this code transform `dist/app2/build/file.js` into `app2/file.js`? Again, I didn't run the code, so I'm not sure my reading is accurate but, if this is the case, that seems pretty random.

Reply:
It shouldn't because it's not checking just the strings, but the index where that string is. It's comparing index 0 of root with index 0 of generated file and so on. I'll write some additional tests anyways just to make sure.

Reply:
This sounds like it's undefined behaviour in general actually. Should it remove as much as in common as possible (in that example, just ```dist``` with the result ```app2/build/file.js```), or should it not remove anything (with the result ```dist/app2/build/file.js```?

Reply:
It should only remove full matches. 

In my opinion, with this option activated, if there's no (full) match with the prefix, I even think the file should be _discarded_ (i.e. no URL produced, and so not accessible through HTTP), otherwise we get 1:n mapping URLs to files, and possible conflicts.

I'm not entirely sure if this could affect Nollup's functioning, but it is consistent with the behaviour of a real web server (files not in public directory are not accessible).

In our situation, I think a file falling outside of the `virtualWrite` prefix would indicate a configuration problem though. I mean, the prefix is expected, in every case, to be a part (parent directory) of `output.dir`. Every produced files will be under `output.dir`. So if a produced file find itself outside of the prefix, to me it seems to mean that the prefix is not good, in every cases... Maybe the best we can do would be to throw an early error when the `virtualWrite` prefix is not a 100% match with `output.dir`'s beginning? Not entirely sure I'm not missing something here, though :-x

That being said, I think it's safe enough to just discard files that don't match the prefix, and let people sort their configuration issues by themselves. Not making it harder for them by dropping random parts of the paths seems the most helpful we can be on our side.

Reply:
That makes sense. Branch updated to reflect the discarding behaviour.